subprojects/core/src/main/java/org/gradle/api/internal/tasks/SimpleWorkResult.java

```diff
@@ -16,14 +16,26 @@
 package org.gradle.api.internal.tasks;
 
 import org.gradle.api.tasks.WorkResult;
+import org.gradle.util.DeprecationLogger;
 
 public class SimpleWorkResult implements WorkResult {
+
     private final boolean didWork;
 
+    public static WorkResult didWork(boolean didWork) {
```

Comment:
I would change `WorkResult` to an abstract class and have it provide this factory method instead of having it on an internal `SimpleWorkResult` class.  My thinking is like this.

If third party plugins were already constructing `SimpleWorkResult` instances even though it was marked as `internal`, I'm assuming that they found it useful to just pass a boolean expression and get a `WorkResult` that matched.  The deprecation below says that they should use either `DID_WORK` or `DID_NO_WORK` wherever they were previously using the `SimpleWorkResult(boolean didWork)`.  However, that's not even advice we are able to apply across our usages internally. Several places in this pull request, we are still using this `didWork()` factory method because the definition of whether a task did work depends on the state of the task which is represented by a boolean expression that is evaluated at runtime.  We should provide this same utility to our third party plugin developers by moving it from an internal class to an external one.

Also, be sure to update the deprecation advice saying that they should use `WorkResult.DID_WORK` or `DID_NO_WORK` or the static factory method `WorkResult.didWork(boolean didWork)`.

Completely a matter of opinion, and not a required change, you might also consider changing the method name to `WorkResult.of(boolean didWork)`.